### PR TITLE
Give the address and an explanation in a different dimensions exception.

### PR DESCRIPTION
**Pull request by joachim-n**
When a merge is impossible because one side has an array where the other has a scalar, the error is just:

> UnexpectedValueException: different dimensions

This should say where in the array structure the array was found.

Also, explaining which side was which is helpfu.

### CODE REVIEW

**Comment by ckressibucher** [comment]
Hi there,

first or all, thanks for helping to improve this library.

I realized that this library is a bit out of date, regarding the supported PHP versions. I plan to release a version 3.0 which declares PHP 7 as a minimum requirement, allowing me to update test dependencies. Can you change the PR to merge into the new branch, v3.0-dev?

Regarding the code:

Yes, the error message is a bit vague. But I would like to keep the public API the same. The additional parameter only makes sense in recursive calls. I would suggest to "split" the method `doMerge` into a public method with the current signature, and a protected method with the additional parameter. The public method invokes the protected method with an empty array as `$address`. Can you adapt your code this way?

File: src/ArrayMerger.php [hunk 1]
```diff
@@ -160,14 +162,20 @@ protected static function mergeRecursively($default, $precedence, $flags)
             return $precedence; // overwrite default by precedence
         }
         if (! ($flags & self::FLAG_ALLOW_SCALAR_TO_ARRAY_CONVERSION)) {
-            throw new \UnexpectedValueException('different dimensions');
+            if (is_array($default)) {
+                $reason = 'default is an array while precedence is a scalar';
```
Comment (ckressibucher) [inline src/ArrayMerger.php]:
```suggestion
                $reason = '`$default` is an array while `$precedence` is a scalar';
```

File: src/ArrayMerger.php [hunk 2]
```diff
@@ -160,14 +162,20 @@ protected static function mergeRecursively($default, $precedence, $flags)
...
+            }
+            else {
+                $reason = 'default is a scalar while precedence is an array';
```
Comment (ckressibucher) [inline src/ArrayMerger.php]:
```suggestion
                $reason = '`$default` is a scalar while `$precedence` is an array';
```

**Review by ckressibucher — COMMENTED** [review]
I think we should make clear that 'default' and 'precedence' are variable names; otherwise the error message may be a bit confusing.

File: src/ArrayMerger.php (hunk 2 above)
Comment (joachim-n) [reply src/ArrayMerger.php]: I've taken that on board, but just quoted 'default', as the value that's caused a problem isn't actually the passed variable $default but a value inside that array.

**Comment by joachim-n** [comment]
Done.

**Pull request merged**